analysis/src/cdecl.rs

```diff
@@ -0,0 +1,478 @@
+                        let checked_dec =
+                            |x: usize| x.checked_sub(1).ok_or(ErrorKind::UnbalancedBrackets);
+                        match right[i] {
+                            CTok::Punct(']' | ')') => nesting += 1,
+                            CTok::Punct('[' | '(') => nesting = checked_dec(nesting)?,
+                            _ => {}
+                        }
+                        if nesting == 0 {
+                            break;
+                        }
+
+                        // Outermost parentheses around the name being declared,
+                        // handled together after both `left` and `right` loops.
+                        if i == 0 && rightmost == CTok::Punct(')') {
+                            break 'right;
+                        }
+
+                        i = checked_dec(i)?;
+                    }
+                    right.split_at(i)
+                };
+
+                match rightmost_group {
+                    [CTok::Punct('['), len @ .., CTok::Punct(']')] => {
+                        ty = CType::Array {
+                            element: Box::new(ty),
+                            len: match len {
+                                [CTok::ValueName(name)] => CArrayLen::Named(name),
+                                [CTok::IntLit(lit)] => CArrayLen::Literal(
+                                    lit.parse().map_err(ErrorKind::InvalidIntLit)?,
+                                ),
+                                _ => return Err(ErrorKind::UnsupportedArrayLength),
+                            },
+                        };
+                    }
+                    [CTok::Punct('('), params @ .., CTok::Punct(')')] => {
+                        if const_qualif.is_some() {
+                            return Err(ErrorKind::Unused("const"));
+                        }
+
+                        let params = match params {
+                            [] => return Err(ErrorKind::Missing("parameters")),
+                            [CTok::Kw("void")] => vec![],
+                            _ => params
+                                .split(|&tok| tok == CTok::Punct(','))
+                                .map(|param| CDecl::parse_inner(CDeclMode::FuncTypeParam, param))
+                                .collect::<Result<_, _>>()?,
+                        };
+                        ty = CType::Func {
+                            ret_ty: Some(ty).filter(|ty| *ty != CType::VOID).map(Box::new),
+                            params,
+                        };
+                    }
+                    _ => return Err(ErrorKind::UnbalancedBrackets),
+                }
+                right = before_rightmost_group;
+            }
+
+            // Outermost parentheses around the name being declared, handled here
+            // to ensure there is nothing else left around them, and can therefore
+            // be cleanly removed.
+            match (left, right) {
```

Comment:
Clippy wants this to be an if-let, because there's only a single case.

Reply:
I'm used to clippy being more annoying about that kind of thing, but here it's probably right, I think I only use `match` in cases like this to not obfuscate the pattern. I should still try the `if let`.

Reply:
> I'm used to clippy being more annoying about that kind of thing

I understand you very well 😅 I resolved this in my branch (https://github.com/Friz64/ash/pull/1#issuecomment-2001889813) so the warning goes away.